src/Control/Concurrent/TMap.hs

```diff
@@ -0,0 +1,137 @@
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE ImplicitPrelude #-}
+
+-- | Definitions for the 'TMap' container.
+--
+-- @since 1.0.0
+module Control.Concurrent.TMap
+  ( -- * TMap
+    TMap (TMap, getTMap),
+
+    -- * Construction
+    empty,
+    emptyIO,
+
+    -- * Destruction
+    toAscList,
+
+    -- * Insertion
+    insert,
+
+    -- * Deletion
+    delete,
+
+    -- * Query
+    lookup,
+
+    -- * Folding
+    ifoldr,
+  )
+where
+
+--------------------------------------------------------------------------------
+
+import Control.Concurrent.STM (STM)
+import Control.Concurrent.STM.TVar (TVar, newTVarIO, readTVar, writeTVar)
+
+import Data.IORef (IORef, newIORef, readIORef, atomicModifyIORef')
+
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+
+import GHC.Conc.Sync (unsafeIOToSTM)
+
+import Prelude hiding (lookup)
+
+-- TMap ------------------------------------------------------------------------
+
+-- | STM-specialized non-blocking 'Map' container.
```

Comment:
Why not use the transactional trie package https://hackage.haskell.org/package/ttrie-0.1.2.2/docs/Control-Concurrent-STM-Map.html?

Reply:
I chose not to use this package because I was trying to avoid using hashed keys to index the map. I've heard of changes to `hashable` internals causing issues for packages before, and using `Ord` lets key comparison can short circuit in the insertion logic.

The `TMap` container provided here should be equivalent map `ttrie` provided otherwise, and can be removed in favor of `ttrie` if you think that would be better.

Reply:
I don't have a strong preference except to use a library when/where it makes sense over hand rolled versions. It sounds like you have a good reason for not using the library though so that's fine with me.